### PR TITLE
Hit And Miss morphological operation

### PR DESCRIPTION
**Pull request by LorenaGdL**
Implementation of Hit And Miss morphological operation

- Based on the code provided in this page: http://opencv-code.com/tutorials/hit-or-miss-transform-in-opencv/ . Minor fixes added to prevent incorrect results due to erosion with all-zeros kernel (refer to this [Q&A post](http://answers.opencv.org/question/28344/working-hit-or-miss-implementation/))

- New documentation refers to upper link, but I plan to modify such tutorial to publish in OpenCV site, showing new functionality of morphologyEx function.

- ocl version not implemented

- If everything is alright and gets merged, I will provide 2.4.x PR

- Code sample showing functionality [here](https://gist.github.com/LorenaGdL/ad3d7fc5ca070a549715)

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
:+1: nice one :)

File: modules/imgproc/src/morph.cpp [hunk 1]
```diff
@@ -1905,6 +1907,22 @@ void cv::morphologyEx( InputArray _src, OutputArray _dst, int op,
         erode( temp, temp, kernel, anchor, iterations, borderType, borderValue );
         dst = temp - src;
         break;
+    case MORPH_HITMISS:
+        CV_Assert(src.type() == CV_8U && src.channels() == 1);
+        k1 = (kernel == 1) / 255;
+        k2 = (kernel == -1) / 255;
+        normalize(src, src, 0, 1, NORM_MINMAX);
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/morph.cpp]: why do you change `src` matrix?

File: modules/imgproc/src/morph.cpp [hunk 2]
```diff
@@ -1905,6 +1907,22 @@ void cv::morphologyEx( InputArray _src, OutputArray _dst, int op,
         erode( temp, temp, kernel, anchor, iterations, borderType, borderValue );
         dst = temp - src;
         break;
+    case MORPH_HITMISS:
+        CV_Assert(src.type() == CV_8U && src.channels() == 1);
```
Comment (ilya-lavrenov) [inline modules/imgproc/src/morph.cpp]: the condition is redundant, `src.type() == CV_8UC1` is enough.

Comment (LorenaGdL) [inline modules/imgproc/src/morph.cpp]: True, I will fix it

File: modules/imgproc/src/morph.cpp (hunk 1 above)
Comment (LorenaGdL) [inline modules/imgproc/src/morph.cpp]: Original author normalized `src` to easily compute its complement for the second erosion (`erode(1-src,...)`). Maybe not the most elegant solution, so I'll remove that normalization and use a `bitwise_not`

**Comment by LorenaGdL** [comment]
@ilya-lavrenov ok, you can take another look

**Comment by vpisarev** [comment]
thanks! :+1:

**Comment by StevenPuttemans** [comment]
@LorenaGdL thank you for the algorithm! Could you provide the same for 2.4?

**Comment by LorenaGdL** [comment]
@StevenPuttemans that was already on my to-do list (check above). I'll do it as soon as I have some spare time, and same applies to the tutorial :)

**Comment by StevenPuttemans** [comment]
:+1:

**Pull request merged**